File: SDE_sheet/Arrays/13_search_in_2D_matrix/solution.go
func searchMatrix(matrix [][]int, target int) bool {
    m, n := len(matrix), len(matrix[0])
    left, right := 0, m*n-1

    for left <= right {
        mid := (left + right) / 2
        row, col := mid/n, mid%n
        if matrix[row][col] == target {
            return true
        }
        if matrix[row][col] > target {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    return false
}

File: SDE_sheet/Arrays/13_search_in_2D_matrix/solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;
        int left = 0, right = m * n - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            int row = mid / n;
            int col = mid % n;
            if (matrix[row][col] == target) {
                return true;
            }
            if (matrix[row][col] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return false;
    }
}
